enhancements/installer/component-selection.md

```diff
@@ -0,0 +1,500 @@
+
+The `inclusionDefault` field (carried through to the CVO) allows an Admin to choose what behavior they want for new
+capabilities that might come as part of a future upgrade (either have the new capability installed during
+the upgrade, or exclude it)
+
+Specifying the same capability as an `include` and an `exclude` is an error condition.
+
+
+Examples of candidate components to be treated as capabilities:
+
+* console
+* imageregistry
+* samples
+* cluster baremetal operator
+* olm/marketplace
+* csi-*
+* insights
+* monitoring
+* ???
+
+
+
+3) Pass the list of filtered annotations to the CVO.  This is distinct from overrides because overrides
+put the cluster in an unsupported state.  Filtered annotations are supported for upgrades.  The filtered
+components will be listed in the [ClusterVersion](https://github.com/openshift/api/blob/4436dc8be01e8dcd8b250e1b32bb0fbd64ba78ac/config/v1/types_cluster_version.go#L35) object:
+
+```yaml
+  spec:
+    capabilities:
+      inclusionDefault: [Include|Exclude]
+      exclude:
+      - console
+      - samples
+      include:
+      - registry
+```
+
+The CVO will filter out(not apply/reconcile) resources that are annotated with key/value
+`capability.openshift.io/name=$exclusions` and listed in the `exclude` section, or are not listed
+and the `default` setting is `exclude`.  In addition any explicitly included capability
+should be added, regardless of the CLUSTER_PROFILE.
+
+If a resource has multiple `capability.openshift.io/name` annotations, then the resource will be filtered if any
+of the annotations on the resource are matched by a configured filter.  This allows a component to define
+a `ServiceMonitor` which is part of both the `Monitoring` capability and the `Foo` capability.  If either
+of those capabilities are excluded, the `ServiceMonitor` should not be created.  (If Monitoring is excluded
+then no ServiceMonitors should be created, if only Foo is excluded, then only Foo's ServiceMonitors should
+be excluded).
+
+Specifying the same capability as an `include` and an `exclude` is an error condition.  The CVO will need to
+implement a validating webhook on ClusterVersion to enforce this.  This avoids the need to define behavior for
+what to do when a user specifies the same capability in both, and update a status field indicating what choice
+the CVO made.
+
+Note: The CVO must wait until the ClusterVersion resource exists (created by the installer) before creating
+any filterable resource(a resource with a capability annotation) to ensure there is no race condition in which
+the CVO starts creating resources that should have been filtered, before it has the filter list.
+
+4) Admin can remove an item from the excluded annotations list, but they cannot add an item to it.  If an
+item is removed, the CVO will apply the previously filtered(skipped) resources to the cluster on the next reconciliation.
```

Comment:
Bumped into with [the tech-preview manifest annotations][1] is that the CVO is not flipping from reconcile-mode to update-mode when it sees the config change.  That means we get the flattened, permuted manifest-task graph, and not [update's ordered, leveled manifest task-node graph][2].  Depending on how many task-nodes are being enabled, there are a few possibilities:

a. Have capability config changes count as "update requests", with a `status.history` entry or some other tracker so a restarted CVO container can figure out when it's continuing a previous, partial update.  Possibly we could hinge this on `Progressing`, without having to poke at `history`.
b. Get all the manifests involved in a capability into a single block, or otherwise structured so that they do not need to care about task-node levels.  Seems tricky for things like monitoring, where folks currently often put their ServiceMonitor at a high level to ensure they come after both monitoring's ServiceMonitor CRD and the component-specific Service being monitored.
c. Ensure components aren't spread across particularly many task-nodes, so that [our 16 graph workers][3] are likely to find the manifests in task-node A that block other manifests in task-node B.  There could be a few rounds of noise here, as a worker racing through task-node B might fail the blocked manifest before its dependency lands in the cluster, but after a few round and some possible noisy alerting, we should eventually have worked through any dependencies.

(a) is the cleanest, but also the most invasive.  (c) is easy, but not the best UX unless we find some clever way to hide  the noise.  As long as we can wedge things into (b)'s mitigation, we can punt on deciding between (a) and (c).

[1]: https://github.com/openshift/enhancements/blob/f74ffe7776f40dbd096b9ca10c27ee7a0a579e58/enhancements/update/cvo-techpreview-manifests.md
[2]: https://github.com/openshift/enhancements/blob/f74ffe7776f40dbd096b9ca10c27ee7a0a579e58/dev-guide/cluster-version-operator/user/reconciliation.md#manifest-graph
[3]: https://github.com/openshift/cluster-version-operator/blob/ede5b83feb38338b31e7549507422f4a76e970e6/pkg/cvo/cvo.go#L361

Reply:
> Isn't any transition from excluded to included the same as install time reconcile mode?

Since way back in openshift/cluster-version-operator#136, the CVO's install mode is basically just shoveling manifests into the cluster as fast as possible without really caring about how they work out, under the assumption that folks want their cluster installed quickly, and they aren't too worried about a bit of noise.  I'm not sure folks will be so complacent about noise after a day-2 capability spec change.

> Isn't any transition from excluded to included the same as install time reconcile mode? I think the CVO just has to update capabilities status to match the spec and continue reconciliation if it were restarted, no?

So breaking down the three CVO modes:

* Install mode.  Parallel task-nodes.  Weaker requirements on the in-cluster resource before the CVO accepts a manifest as "reconciled" and moves on to the next manifest or task-node.
* Reconcile mode.  Parallel task-nodes.  Usual requirements on the in-cluster resource.
* Update mode.  Ordered task-node graph, so [level 50 stuff][1] is all resolved before [level 90 stuff][2], etc.  Usual requirements on the in-cluster resource.

If a capabilities change enables several task nodes at different levels, and there's some dependency (e.g. [an earlier monitoring task-node with a ServiceMonitor CRD][1], and a later task-node that attempts to install some [ServiceMonitor for monitoring the CVO][2]), both install mode and reconcile mode could experience some noise, if the CVO randomly decided to pick the task node with the CVO ServiceMonitor manifest to reconcile before it attempted the CRD manifest.  But we'd eventually come out the other side and all would be fine.

If we use update mode for the transition, [the level 50 CRD][1] would always complete before [the level 90 consumer][2], but we'd need something like a `status.history` entry so the CVO would know "hey, I should be in update mode" in case the pod was restarted before it finished a full reconcile after the spec bump.

[1]: https://github.com/openshift/cluster-monitoring-operator/blob/b07ad32026ffd600f0a6710addade6f0f1808f07/manifests/0000_50_cluster-monitoring-operator_00_0servicemonitor-custom-resource-definition.yaml
[2]: https://github.com/openshift/cluster-version-operator/blob/3db74ebb012838ffdd776ca65210e4d6d66b7472/install/0000_90_cluster-version-operator_02_servicemonitor.yaml

Reply:
Given that only an admin could push something from excluded to included I'm not too worried about noise during that transition. I may be more cavalier than others however.